api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -300,101 +299,88 @@ private DataSetRequest generateDataSetRequestFromPreviewRequest(
-                              } catch (NumberFormatException ex) {
-                                queryValues.add("");
-                              }
-                            });
-                    previewValue.setQueryValue(queryValues);
-                  }
-                });
-        break;
-      } catch (Exception ex) {
-        if ((ex.getCause() != null
-            && ex.getCause().getMessage() != null
-            && ex.getCause().getMessage().contains("Read timed out"))) {
-          rowsRequested = (rowsRequested / 2);
-          if (rowsRequested == 0) {
-            throw new GatewayTimeoutException(
-                "Timeout while querying the CDR to pull preview information.");
-          }
-          retry++;
-        } else {
-          throw ex;
-        }
+    TableResult queryResponse;
+    try {
+      String query = originalQuery.concat(" LIMIT " + NO_OF_PREVIEW_ROWS);
+
+      queryJobConfiguration = queryJobConfiguration.toBuilder().setQuery(query).build();
+
+      /* Google appengine has a 60 second timeout, we want to make sure this endpoint completes
+       * before that limit is exceeded, or we get a 500 error with the following type:
+       * com.google.apphosting.runtime.HardDeadlineExceededError
+       * See https://cloud.google.com/appengine/articles/deadlineexceedederrors for details
+       */
+      queryResponse =
+          bigQueryService.executeQuery(
+              bigQueryService.filterBigQueryConfig(queryJobConfiguration),
+              APP_ENGINE_HARD_TIMEOUT_MSEC_MINUS_FIVE_SEC);
+    } catch (Exception ex) {
+      if ((ex.getCause() != null
+          && ex.getCause().getMessage() != null
+          && ex.getCause().getMessage().contains("Read timed out"))) {
+        throw new GatewayTimeoutException(
+            "Timeout while querying the CDR to pull preview information.");
+      } else {
+        throw ex;
       }
-    } while (retry < PREVIEW_RETRY_LIMIT);
-
-    if (retry == PREVIEW_RETRY_LIMIT) {
-      throw new GatewayTimeoutException(
-          "Timeout while querying the CDR to pull preview information.");
     }
+    queryResponse
+        .getSchema()
+        .getFields()
+        .forEach(
+            fields -> {
+              valuePreviewList.add(new DataSetPreviewValueList().value(fields.getName()));
+            });
+
+    queryResponse
+        .getValues()
+        .forEach(
+            fieldValueList -> {
+              IntStream.range(0, fieldValueList.size())
+                  .forEach(
+                      columnNumber -> {
+                        try {
+                          valuePreviewList
+                              .get(columnNumber)
+                              .addQueryValueItem(
+                                  fieldValueList.get(columnNumber).getValue().toString());
+                        } catch (NullPointerException ex) {
+                          log.severe(
```

Comment:
So in this situation, why do we keep going? It seems like the user will get a query with fewer values than requested? I don't see how the client is going to know anything is amiss.

Reply:
We've determined that null is a valid response for many of these columns where there is no data. That said, I don't really like catching a null pointer. So I can poke at trying to fix that. 

Whether we keep going or not here is a product question though, and I would argue is outside the scope of this ticket.